File: frontend/src/lib/subscription.ts
export type SubscriptionTier = 'FREE' | 'PAY_PER_ANALYSIS' | 'PRO_LITE' | 'PRO';

export interface PlanFeature {
  name: string;
  included: boolean;
}

export interface PlanConfig {
  tier: SubscriptionTier;
  name: string;
  price: string;
  interval: string;
  description: string;
  monthlyLimit: number;
  aiModel: string;
  pipelineType: string;
  popular?: boolean;
  features: PlanFeature[];
}

export const SUBSCRIPTION_PLANS: Record<SubscriptionTier, PlanConfig> = {
  FREE: {
    tier: 'FREE',
    name: 'Free',
    price: '₹0',
    interval: '/ month',
    description: 'Perfect for exploring Apex Intel before upgrading.',
    monthlyLimit: 2,
    aiModel: 'Gemini 2.5 Flash Lite',
    pipelineType: 'Optimized 3-Agent Pipeline',
    features: [
      { name: '2 analyses per month', included: true },
      { name: 'Uses previous-generation AI models', included: true },
      { name: 'Optimized 3-Agent Pipeline', included: true },
      { name: 'Basic company briefing', included: true },
      { name: 'Contradiction detection', included: false },
      { name: 'Hidden assumptions validation', included: false },
      { name: 'Detailed Investment Score Breakdown', included: false },
    ],
  },
  PAY_PER_ANALYSIS: {
    tier: 'PAY_PER_ANALYSIS',
    name: 'Pay-Per-Analysis',
    price: '₹29',
    interval: '/ report',
    description: 'Need only one report? Purchase a single premium analysis without a subscription.',
    monthlyLimit: 0,
    aiModel: 'Gemini 2.5 Flash',
    pipelineType: 'Full 9-Agent Pipeline',
    features: [
      { name: 'One-time purchase', included: true },
      { name: 'Credits never expire', included: true },
      { name: 'No subscription required', included: true },
      { name: 'Premium AI Model Access', included: true },
      { name: 'Complete 9-Agent Due Diligence', included: true },
      { name: 'Detailed Investment Score Breakdown', included: true },
      { name: 'Priority Founder Support', included: false },
    ],
  },
  PRO_LITE: {
    tier: 'PRO_LITE',
    name: 'Pro Lite',
    price: '₹45',
    interval: '/ month',
    description: 'Ideal for students, founders, and regular startup research.',
    monthlyLimit: 5,
    aiModel: 'Gemini 2.5 Flash',
    pipelineType: 'Standard 5-Agent Pipeline',
    features: [
      { name: '5 analyses per month', included: true },
      { name: 'Premium AI Model Access', included: true },
      { name: 'Standard 5-Agent Pipeline', included: true },
      { name: 'Full Investment Report', included: true },
      { name: 'Contradiction detection', included: false },
      { name: 'Hidden assumptions validation', included: false },
      { name: 'Detailed Investment Score Breakdown', included: false },
    ],
  },
  PRO: {
    tier: 'PRO',
    name: 'Pro',
    price: '₹56',
    interval: '/ month',
    popular: true,
    description: 'Complete institutional-grade AI due diligence for investors and power users.',
    monthlyLimit: 10,
    aiModel: 'Gemini 2.5 Flash',
    pipelineType: 'Full 9-Agent Pipeline',
    features: [
      { name: '10 analyses per month', included: true },
      { name: 'Premium AI Model Access', included: true },
      { name: 'Complete 9-Agent Due Diligence', included: true },
      { name: 'Full Investment Report', included: true },
      { name: 'Contradiction detection engine', included: true },
      { name: 'Hidden assumptions validation', included: true },
      { name: 'Detailed Investment Score Breakdown', included: true },
      { name: 'Priority Founder Support', included: true },
    ],
  },
};

export const getPlanConfig = (tier?: string | null): PlanConfig => {
  if (!tier) return SUBSCRIPTION_PLANS.FREE;
  const config = SUBSCRIPTION_PLANS[tier as SubscriptionTier];
  return config || SUBSCRIPTION_PLANS.FREE;
};

export const getPlanRank = (tier: string): number => {
  const ranks: Record<string, number> = {
    FREE: 0,
    PAY_PER_ANALYSIS: 1, // Treat as side-grade or level 1
    PRO_LITE: 2,
    PRO: 3
  };
  return ranks[tier] ?? 0;
};

File: backend/core/subscription.py
from typing import Dict, List, Optional
from pydantic import BaseModel
from backend.core.enums import SubscriptionTier

class PlanFeature(BaseModel):
    name: str
    included: bool

class PlanConfig(BaseModel):
    tier: SubscriptionTier
    name: str
    price: int  # Stored as integer value (e.g. 56)
    currency: str = "₹"
    interval: str = "/ month"
    description: str
    monthly_limit: int
    ai_model: str
    pipeline_type: str
    popular: bool = False
    features: List[PlanFeature]
    rank: int
    stripe_product_env: Optional[str] = None
    stripe_price_env: Optional[str] = None

# Single Source of Truth for the Backend
SUBSCRIPTION_PLANS: Dict[SubscriptionTier, PlanConfig] = {
    SubscriptionTier.FREE: PlanConfig(
        tier=SubscriptionTier.FREE,
        name="Free",
        price=0,
        interval="/ month",
        description="Perfect for exploring Apex Intel before upgrading.",
        monthly_limit=2,
        ai_model="Gemini 2.5 Flash Lite",
        pipeline_type="Optimized 3-Agent Pipeline",
        rank=0,
        features=[
            PlanFeature(name="2 analyses per month", included=True),
            PlanFeature(name="Uses previous-generation AI models", included=True),
            PlanFeature(name="Optimized 3-Agent Pipeline", included=True),
            PlanFeature(name="Basic company briefing", included=True),
            PlanFeature(name="Contradiction detection", included=False),
            PlanFeature(name="Hidden assumptions validation", included=False),
            PlanFeature(name="Detailed Investment Score Breakdown", included=False),
        ]
    ),
    SubscriptionTier.PAY_PER_ANALYSIS: PlanConfig(
        tier=SubscriptionTier.PAY_PER_ANALYSIS,
        name="Pay-Per-Analysis",
        price=29,
        interval="/ report",
        description="Need only one report? Purchase a single premium analysis without a subscription.",
        monthly_limit=0,
        ai_model="Gemini 2.5 Flash",
        pipeline_type="Full 9-Agent Pipeline",
        rank=1,
        stripe_price_env="STRIPE_PRICE_CREDIT",
        features=[
            PlanFeature(name="One-time purchase", included=True),
            PlanFeature(name="Credits never expire", included=True),
            PlanFeature(name="No subscription required", included=True),
            PlanFeature(name="Premium AI Model Access", included=True),
            PlanFeature(name="Complete 9-Agent Due Diligence", included=True),
            PlanFeature(name="Detailed Investment Score Breakdown", included=True),
            PlanFeature(name="Priority Founder Support", included=False),
        ]
    ),
    SubscriptionTier.PRO_LITE: PlanConfig(
        tier=SubscriptionTier.PRO_LITE,
        name="Pro Lite",
        price=45,
        interval="/ month",
        description="Ideal for students, founders, and regular startup research.",
        monthly_limit=5,
        ai_model="Gemini 2.5 Flash",
        pipeline_type="Standard 5-Agent Pipeline",
        rank=2,
        stripe_price_env="STRIPE_PRICE_PRO_LITE",
        features=[
            PlanFeature(name="5 analyses per month", included=True),
            PlanFeature(name="Premium AI Model Access", included=True),
            PlanFeature(name="Standard 5-Agent Pipeline", included=True),
            PlanFeature(name="Full Investment Report", included=True),
            PlanFeature(name="Contradiction detection", included=False),
            PlanFeature(name="Hidden assumptions validation", included=False),
            PlanFeature(name="Detailed Investment Score Breakdown", included=False),
        ]
    ),
    SubscriptionTier.PRO: PlanConfig(
        tier=SubscriptionTier.PRO,
        name="Pro",
        price=56,
        interval="/ month",
        description="Complete institutional-grade AI due diligence for investors and power users.",
        monthly_limit=10,
        ai_model="Gemini 2.5 Flash",
        pipeline_type="Full 9-Agent Pipeline",
        popular=True,
        rank=3,
        stripe_price_env="STRIPE_PRICE_PRO",
        features=[
            PlanFeature(name="10 analyses per month", included=True),
            PlanFeature(name="Premium AI Model Access", included=True),
            PlanFeature(name="Complete 9-Agent Due Diligence", included=True),
            PlanFeature(name="Full Investment Report", included=True),
            PlanFeature(name="Contradiction detection engine", included=True),
            PlanFeature(name="Hidden assumptions validation", included=True),
            PlanFeature(name="Detailed Investment Score Breakdown", included=True),
            PlanFeature(name="Priority Founder Support", included=True),
        ]
    )
}

def get_plan_config(tier: SubscriptionTier) -> PlanConfig:
    """Returns the strongly-typed plan configuration."""
    return SUBSCRIPTION_PLANS.get(tier, SUBSCRIPTION_PLANS[SubscriptionTier.FREE])

def get_stripe_price_id(tier: SubscriptionTier) -> Optional[str]:
    """Resolves the Stripe Price ID from settings dynamically based on the plan's env mapping."""
    from backend.config.settings import settings
    config = get_plan_config(tier)
    if not config.stripe_price_env:
        return None
    return getattr(settings, config.stripe_price_env, None)
